include/util/filtered_graph.hpp

```diff
@@ -0,0 +1,165 @@
+    {
+        for (const auto edge : GetAdjacentEdgeRange(from))
+        {
+            if (to == GetTarget(edge))
+            {
+                return edge;
+            }
+        }
+        return SPECIAL_EDGEID;
+    }
+
+    template <typename FilterFunction>
+    EdgeIterator
+    FindSmallestEdge(const NodeIterator from, const NodeIterator to, FilterFunction &&filter) const
+    {
+        static_assert(traits::HasDataMember<typename Graph::EdgeArrayEntry>::value,
+                      "Filtering on .data not possible without .data member attribute");
+
+        EdgeIterator smallest_edge = SPECIAL_EDGEID;
+        EdgeWeight smallest_weight = INVALID_EDGE_WEIGHT;
+        for (auto edge : GetAdjacentEdgeRange(from))
+        {
+            const NodeID target = GetTarget(edge);
+            const auto &data = GetEdgeData(edge);
+            if (target == to && data.weight < smallest_weight &&
+                std::forward<FilterFunction>(filter)(data))
+            {
+                smallest_edge = edge;
+                smallest_weight = data.weight;
+            }
+        }
+        return smallest_edge;
+    }
+
+    EdgeIterator FindEdgeInEitherDirection(const NodeIterator from, const NodeIterator to) const
+    {
+        EdgeIterator tmp = FindEdge(from, to);
+        return (SPECIAL_NODEID != tmp ? tmp : FindEdge(to, from));
+    }
+
+    EdgeIterator
+    FindEdgeIndicateIfReverse(const NodeIterator from, const NodeIterator to, bool &result) const
+    {
+        EdgeIterator current_iterator = FindEdge(from, to);
+        if (SPECIAL_NODEID == current_iterator)
+        {
+            current_iterator = FindEdge(to, from);
+            if (SPECIAL_NODEID != current_iterator)
+            {
+                result = true;
+            }
+        }
+        return current_iterator;
+    }
+
+    FilteredGraphImpl() = default;
+
+    FilteredGraphImpl(Graph graph, Vector<bool> edge_filter_)
+        : graph(std::move(graph)), edge_filter(std::move(edge_filter_))
+    {
+        BOOST_ASSERT(edge_filter.empty() || edge_filter.size() == graph.GetNumberOfEdges());
+    }
+
+    // Takes a graph and a function that maps EdgeID to true
+    // if the edge should be included in the graph.
+    template <typename Pred>
+    FilteredGraphImpl(Graph graph, Pred filter)
+        : graph(std::move(graph)), edge_filter(graph.GetNumberOfEdges())
+    {
+        auto edge_ids = util::irange<EdgeID>(0, graph.GetNumberOfEdges());
+        std::transform(edge_ids.begin(), edge_ids.end(), edge_filter.begin(), filter);
+    }
+
+    void Renumber(const std::vector<NodeID> &old_to_new_node)
+    {
+        graph.Renumber(old_to_new_node);
+        // FIXME the edge filter needs to be renumbered with a different permutation
+        // util::inplacePermutation(edge_filter.begin(), edge_filter.end(), old_to_new_node);
+    }
+
+  private:
+    Graph graph;
+    Vector<bool> edge_filter;
```

Comment:
I think that memory usage can be minimized 8x times (depending on compiler and platform). Just using of bit vector. You need 1 bit to store bool value. Math to access it very simple:
```cpp
bool isFilter(EdgeID eid)
{
    return edge_filter[eid / 8] & (1 << (eid % 8));
}
```

instead `eid / 8` can be used `eid >> 3`. So it would be fast and use `EdgeNumber / 8` bytes of memory

Reply:
@deniskoronchik internally [`Vector<bool> `](https://github.com/Project-OSRM/osrm-backend/blob/993faba545e90a4ae0bd9027119f45b9635176c5/include/util/vector_view.hpp#L142-L206)
```
(gdb) ptype QueryGraph::edge_filter
type = class osrm::util::vector_view<bool> {
  private:
    unsigned int *m_ptr;
    std::size_t m_size;
    static const std::size_t UNSIGNED_BITS;
...
}
```

ATM it is stored in a file as 8 bits per value, but there is a not-yet merged commit 69d53ad6db64207f565dda7c920b9dec5834d5cc to reduce file sizes